### PR TITLE
Redesign shop api (PLEASE TEST AND REVIEW)

### LINKED ISSUE #71: Delete/View functions retrieval of item is wrong
image

get(index) gets relative to list position and not the actual object's internal id.

### LINKED ISSUE #99: view part does not list parts

### LINKED ISSUE #122: [PE-D][Tester C] Technician doesn't show up after adding
1. Run `addtechnician`
2. The screen loads to technician tab but it is empty.

-------------
Labels: `type.FunctionalityBug` `severity.High`
original: zm-l/ped#2

### LINKED ISSUE #125: [PE-D][Tester B] [addtechnician] NullPointerException after running addtechnician command
Command:
```
addtechnician n/Tester2 p/12345678 e/test@test.com a/123, test address t/121 t/212
```

Stacktrace:
```
Result: New technician added: Tester2;
Phone: 12345678;
Email: test@test.com
Address: 123, test address; Tags: [121][212]
Exception in thread "JavaFX Application Thread" java.lang.NullPointerException
        at seedu.address.model.mapping.TechnicianDataMap.getTechnicianServices(TechnicianDataMap.java:79)
        at seedu.address.ui.TechnicianListPanel$TechnicianListViewCell.updateItem(TechnicianListPanel.java:49)
        at seedu.address.ui.TechnicianListPanel$TechnicianListViewCell.updateItem(TechnicianListPanel.java:39)
        at javafx.scene.control.ListCell.updateItem(ListCell.java:478)
        at javafx.scene.control.ListCell.indexChanged(ListCell.java:337)
        at javafx.scene.control.IndexedCell.updateIndex(IndexedCell.java:120)
        at javafx.scene.control.skin.VirtualFlow.setCellIndex(VirtualFlow.java:1708)
        at javafx.scene.control.skin.VirtualFlow.addLeadingCells(VirtualFlow.java:1947)
        at javafx.scene.control.skin.VirtualFlow.layoutChildren(VirtualFlow.java:1250)
        at javafx.scene.control.skin.VirtualFlow$5.invalidated(VirtualFlow.java:837)
        at javafx.beans.property.IntegerPropertyBase.markInvalid(IntegerPropertyBase.java:113)
        at javafx.beans.property.IntegerPropertyBase.set(IntegerPropertyBase.java:148)
        at javafx.scene.control.skin.VirtualFlow.setCellCount(VirtualFlow.java:855)
        at javafx.scene.control.skin.ListViewSkin.updateItemCount(ListViewSkin.java:333)
        at javafx.scene.control.skin.VirtualContainerBase.checkState(VirtualContainerBase.java:170)
        at javafx.scene.control.skin.VirtualContainerBase.layoutChildren(VirtualContainerBase.java:145)
        at javafx.scene.control.skin.ListViewSkin.layoutChildren(ListViewSkin.java:276)
        at javafx.scene.control.Control.layoutChildren(Control.java:601)
        at javafx.scene.Parent.layout(Parent.java:1204)
        at javafx.scene.Parent.layout(Parent.java:1211)
        at javafx.scene.Parent.layout(Parent.java:1211)
        at javafx.scene.Parent.layout(Parent.java:1211)
        at javafx.scene.Parent.layout(Parent.java:1211)
        at javafx.scene.Parent.layout(Parent.java:1211)
        at javafx.scene.Parent.layout(Parent.java:1211)
        at javafx.scene.Parent.layout(Parent.java:1211)
        at javafx.scene.Parent.layout(Parent.java:1211)
        at javafx.scene.Parent.layout(Parent.java:1211)
        at javafx.scene.Parent.layout(Parent.java:1211)
        at javafx.scene.Parent.layout(Parent.java:1211)
        at javafx.scene.Scene.doLayoutPass(Scene.java:576)
        at javafx.scene.Scene$ScenePulseListener.pulse(Scene.java:2482)
        at com.sun.javafx.tk.Toolkit.lambda$runPulse$2(Toolkit.java:412)
        at java.base/java.security.AccessController.doPrivileged(Native Method)
        at com.sun.javafx.tk.Toolkit.runPulse(Toolkit.java:411)
        at com.sun.javafx.tk.Toolkit.firePulse(Toolkit.java:438)
        at com.sun.javafx.tk.quantum.QuantumToolkit.pulse(QuantumToolkit.java:519)
        at com.sun.javafx.tk.quantum.QuantumToolkit.pulse(QuantumToolkit.java:499)
        at com.sun.javafx.tk.quantum.QuantumToolkit.pulseFromQueue(QuantumToolkit.java:492)
        at com.sun.javafx.tk.quantum.QuantumToolkit.lambda$runToolkit$11(QuantumToolkit.java:320)
        at com.sun.glass.ui.InvokeLaterDispatcher$Future.run(InvokeLaterDispatcher.java:96)
        at com.sun.glass.ui.win.WinApplication._runLoop(Native Method)
        at com.sun.glass.ui.win.WinApplication.lambda$runLoop$3(WinApplication.java:174)
        at java.base/java.lang.Thread.run(Thread.java:834)
```

Could be related to UI layer as the technician list freezes after the command.

-------------
Labels: `severity.Medium` `type.FunctionalityBug`
original: c0j0s/ped#3

### LINKED ISSUE #129: [PE-D][Tester C] Negative value for quantity in addservicepart
With negative value the part quantity will increase if you toggle to service when running the command.
Initially horn was -1, after the command it becomes 101.

-------------
Labels: `type.FunctionalityBug` `severity.High`
original: zm-l/ped#7

### LINKED ISSUE #132: [PE-D][Tester B] [editvehicle] Unhandled NullPointerException after running editvehicle command
The command is successful but throws NullPointerException silently.

Stacktrace:
```
Exception in thread "JavaFX Application Thread" java.lang.NullPointerException
        at seedu.address.ui.VehicleCard.<init>(VehicleCard.java:70)
        at seedu.address.ui.VehicleListPanel$VehicleListViewCell.updateItem(VehicleListPanel.java:48)
        at seedu.address.ui.VehicleListPanel$VehicleListViewCell.updateItem(VehicleListPanel.java:39)
        at javafx.scene.control.ListCell.updateItem(ListCell.java:478)
        at javafx.scene.control.ListCell.lambda$new$2(ListCell.java:168)
        at javafx.collections.WeakListChangeListener.onChanged(WeakListChangeListener.java:88)
        at com.sun.javafx.collections.ListListenerHelper$Generic.fireValueChangedEvent(ListListenerHelper.java:329)
        at com.sun.javafx.collections.ListListenerHelper.fireValueChangedEvent(ListListenerHelper.java:73)
        at javafx.collections.ObservableListBase.fireChange(ObservableListBase.java:233)
        at javafx.collections.transformation.SortedList.sourceChanged(SortedList.java:114)
        at javafx.collections.transformation.TransformationList.lambda$getListener$0(TransformationList.java:106)
        at javafx.collections.WeakListChangeListener.onChanged(WeakListChangeListener.java:88)
        at com.sun.javafx.collections.ListListenerHelper$SingleChange.fireValueChangedEvent(ListListenerHelper.java:164)
        at com.sun.javafx.collections.ListListenerHelper.fireValueChangedEvent(ListListenerHelper.java:73)
        at javafx.collections.ObservableListBase.fireChange(ObservableListBase.java:233)
        at javafx.collections.ListChangeBuilder.commit(ListChangeBuilder.java:482)
        at javafx.collections.ListChangeBuilder.endChange(ListChangeBuilder.java:541)
        at javafx.collections.ObservableListBase.endChange(ObservableListBase.java:205)
        at javafx.collections.transformation.FilteredList.sourceChanged(FilteredList.java:147)
        at javafx.collections.transformation.TransformationList.lambda$getListener$0(TransformationList.java:106)
        at javafx.collections.WeakListChangeListener.onChanged(WeakListChangeListener.java:88)
        at com.sun.javafx.collections.ListListenerHelper$SingleChange.fireValueChangedEvent(ListListenerHelper.java:164)
        at com.sun.javafx.collections.ListListenerHelper.fireValueChangedEvent(ListListenerHelper.java:73)
        at javafx.collections.ObservableListBase.fireChange(ObservableListBase.java:233)
        at javafx.collections.FXCollections$UnmodifiableObservableListImpl.lambda$new$0(FXCollections.java:955)
        at javafx.collections.WeakListChangeListener.onChanged(WeakListChangeListener.java:88)
        at com.sun.javafx.collections.ListListenerHelper$SingleChange.fireValueChangedEvent(ListListenerHelper.java:164)
        at com.sun.javafx.collections.ListListenerHelper.fireValueChangedEvent(ListListenerHelper.java:73)
        at javafx.collections.ObservableListBase.fireChange(ObservableListBase.java:233)
        at javafx.collections.ListChangeBuilder.commit(ListChangeBuilder.java:482)
        at javafx.collections.ListChangeBuilder.endChange(ListChangeBuilder.java:541)
        at javafx.collections.ObservableListBase.endChange(ObservableListBase.java:205)
        at javafx.collections.ModifiableObservableListBase.set(ModifiableObservableListBase.java:163)
        at seedu.address.model.service.UniqueVehicleList.setVehicle(UniqueVehicleList.java:71)
        at seedu.address.model.entity.shop.Shop.setVehicle(Shop.java:483)
        at seedu.address.model.ModelManager.setVehicle(ModelManager.java:342)
        at seedu.address.logic.commands.EditVehicleCommand.executeUndoableCommand(EditVehicleCommand.java:98)
        at seedu.address.logic.commands.RedoableCommand.execute(RedoableCommand.java:60)
        at seedu.address.logic.LogicManager.execute(LogicManager.java:61)
        at seedu.address.ui.MainWindow.executeCommand(MainWindow.java:327)
        at seedu.address.ui.CommandBox.handleCommandEntered(CommandBox.java:45)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
        at com.sun.javafx.reflect.Trampoline.invoke(MethodUtil.java:76)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
        at com.sun.javafx.reflect.MethodUtil.invoke(MethodUtil.java:273)
        at com.sun.javafx.fxml.MethodHelper.invoke(MethodHelper.java:83)
        at javafx.fxml.FXMLLoader$MethodHandler.invoke(FXMLLoader.java:1784)
        at javafx.fxml.FXMLLoader$ControllerMethodEventHandler.handle(FXMLLoader.java:1670)
        at com.sun.javafx.event.CompositeEventHandler.dispatchBubblingEvent(CompositeEventHandler.java:86)
        at com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(EventHandlerManager.java:238)
        at com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(EventHandlerManager.java:191)
        at com.sun.javafx.event.CompositeEventDispatcher.dispatchBubblingEvent(CompositeEventDispatcher.java:59)
        at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:58)
        at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
        at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56)
        at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
        at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56)
        at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
        at com.sun.javafx.event.EventUtil.fireEventImpl(EventUtil.java:74)
        at com.sun.javafx.event.EventUtil.fireEvent(EventUtil.java:49)
        at javafx.event.Event.fireEvent(Event.java:198)
        at javafx.scene.Node.fireEvent(Node.java:8879)
        at com.sun.javafx.scene.control.behavior.TextFieldBehavior.fire(TextFieldBehavior.java:184)
        at com.sun.javafx.scene.control.behavior.TextInputControlBehavior.lambda$keyMapping$62(TextInputControlBehavior.java:330)
        at com.sun.javafx.scene.control.inputmap.InputMap.handle(InputMap.java:274)
        at com.sun.javafx.event.CompositeEventHandler$NormalEventHandlerRecord.handleBubblingEvent(CompositeEventHandler.java:218)
        at com.sun.javafx.event.CompositeEventHandler.dispatchBubblingEvent(CompositeEventHandler.java:80)
        at com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(EventHandlerManager.java:238)
        at com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(EventHandlerManager.java:191)
        at com.sun.javafx.event.CompositeEventDispatcher.dispatchBubblingEvent(CompositeEventDispatcher.java:59)
        at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:58)
        at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
        at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56)
        at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
        at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56)
        at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
        at com.sun.javafx.event.EventUtil.fireEventImpl(EventUtil.java:74)
        at com.sun.javafx.event.EventUtil.fireEvent(EventUtil.java:54)
        at javafx.event.Event.fireEvent(Event.java:198)
        at javafx.scene.Scene$KeyHandler.process(Scene.java:4058)
        at javafx.scene.Scene$KeyHandler.access$1500(Scene.java:4004)
        at javafx.scene.Scene.processKeyEvent(Scene.java:2121)
        at javafx.scene.Scene$ScenePeerListener.keyEvent(Scene.java:2595)
        at com.sun.javafx.tk.quantum.GlassViewEventHandler$KeyEventNotification.run(GlassViewEventHandler.java:217)
        at com.sun.javafx.tk.quantum.GlassViewEventHandler$KeyEventNotification.run(GlassViewEventHandler.java:149)
        at java.base/java.security.AccessController.doPrivileged(Native Method)
        at com.sun.javafx.tk.quantum.GlassViewEventHandler.lambda$handleKeyEvent$1(GlassViewEventHandler.java:248)
        at com.sun.javafx.tk.quantum.QuantumToolkit.runWithoutRenderLock(QuantumToolkit.java:390)
        at com.sun.javafx.tk.quantum.GlassViewEventHandler.handleKeyEvent(GlassViewEventHandler.java:247)
        at com.sun.glass.ui.View.handleKeyEvent(View.java:547)
        at com.sun.glass.ui.View.notifyKey(View.java:971)
        at com.sun.glass.ui.win.WinApplication._runLoop(Native Method)
        at com.sun.glass.ui.win.WinApplication.lambda$runLoop$3(WinApplication.java:174)
        at java.base/java.lang.Thread.run(Thread.java:834)
```

-------------
Labels: `severity.Medium` `type.FunctionalityBug`
original: c0j0s/ped#6

### PR DESCRIPTION
**Pull request by ginloy**
### Purpose
- Separate backend logic from parsing and command classes (Minimize number of classes to debug)
- Automate everything via one class (Cascading edits/deletes, adding, relations between entities)
- Isolate backend bugs to a single class
- Isolate ID generation
### How it works
- Shop now has methods to add, delete, modify etc
- Methods will handle all checks, cascading effects, error messages
- Ideally it cannot throw unchecked exceptions
- Any modifications to shop and its entities should be done through the API (viewing is ok)
- DONT USE IdGenerator outside of Shop.java (and adding from file)
#### Related issues
- resolve #99 
- should resolve #71
- should resolve #122 
- should resolve #125 
- should resolve #129 
- should resolve #132
##### Note
- Commented out some of @junlee1991 Junit tests to pass CI, please see Shop API `addX()` to see how to modify

### CODE REVIEW

**Review by kimberlybp — CHANGES_REQUESTED** [review]
Remember to remove unused/deprecated classes if this PR removes them (e.g. UniqueList)

File: src/test/java/seedu/address/logic/commands/UndoCommandTest.java
Comment (junlee1991) [inline src/test/java/seedu/address/logic/commands/UndoCommandTest.java:1]: my test got commented out :(

File: src/main/java/seedu/address/model/entity/shop/exception/InsufficientPartException.java
Comment (junlee1991) [inline src/main/java/seedu/address/model/entity/shop/exception/InsufficientPartException.java:1]: My test got commented out :(

File: src/main/java/seedu/address/model/entity/shop/Shop.java [hunk 1]
```diff
@@ -50,500 +74,1012 @@ public Shop(ReadOnlyShop toBeCopied) {
...
+     * Checks if the vehicle exists in the shop.
      *
-     * @param appointmentId ID of appointment
+     * @param id id of the vehicle
+     * @return true if the vehicle exists, false otherwise
      */
-    public boolean hasAppointment(int appointmentId) {
-        return this.getAppointmentList().stream().anyMatch(a -> a.getId() == appointmentId);
+    public boolean hasVehicle(int id) {
+        return this.vehicles.stream().anyMatch(v -> v.getId() == id);
```
[the remark this answers is not in the document]
Comment (ginloy) [reply src/main/java/seedu/address/model/entity/shop/Shop.java:221]: Theoretically it is impossible to add a duplicate plate number unless you extract the the customer out from the filtered list or you abuse `resetData() || overrideData()`. This method is needed for `viewX` which doesn't have access to attributes other than id and the jsonStorage methods which the user should not be able to call.

File: src/main/java/seedu/address/model/entity/shop/Shop.java [hunk 2]
```diff
@@ -50,500 +74,1012 @@ public Shop(ReadOnlyShop toBeCopied) {
...
+     * Checks if the customer exists in the shop.
      *
-     * @param appointment Appointment to check
+     * @param id id of the customer
+     * @return true if the customer exists, false otherwise
      */
-    public boolean hasAppointment(Appointment appointment) {
-        return appointments.contains(appointment);
+    public boolean hasCustomer(int id) {
+        return this.customers.stream().anyMatch(c -> c.getId() == id);
```
[the remark this answers is not in the document]
Comment (ginloy) [reply src/main/java/seedu/address/model/entity/shop/Shop.java:211]: see above

File: src/main/java/seedu/address/storage/JsonSerializableShop.java [hunk 3]
```diff
@@ -81,59 +83,72 @@ public JsonSerializableShop(ReadOnlyShop source) {
      */
     public Shop toModelType() throws IllegalValueException {
         Shop shop = new Shop();
+        List<Customer> customerLst = new ArrayList<>();
```
[the remark this answers is not in the document]
Comment (ginloy) [reply src/main/java/seedu/address/storage/JsonSerializableShop.java:86]: fixed

File: src/main/java/seedu/address/ui/ServiceDetailsPanel.java [hunk 4]
```diff
@@ -83,7 +82,7 @@ private void update(Service service, ServiceDataMap dataMap) {
         vehicleAssignedTo.getChildren().add(vehicleLabel);
 
         List<Technician> technicians = dataMap.getServiceTechnicians(service);
-        PartMap partsRequired = service.getRequiredParts();
+        var partsRequired = service.getRequiredParts();
```
[the remark this answers is not in the document]
Comment (ginloy) [reply src/main/java/seedu/address/ui/ServiceDetailsPanel.java:85]: fixed

File: src/main/java/seedu/address/model/entity/shop/Shop.java [hunk 5]
```diff
@@ -50,500 +74,1012 @@ public Shop(ReadOnlyShop toBeCopied) {
...
 
-    private Optional<Appointment> getAppointment(int appointmentId) {
-        return this.getAppointmentList().stream()
-            .filter(a -> a.getId() == appointmentId)
-            .findFirst();
+    /**
+     * Replaces the existing data of this {@code shop} with {@code newData}.
+     * This should only be called when loading data from file
+     */
+    public void overrideData(ObservableList<Customer> customers,
```
[the remark this answers is not in the document]
Comment (ginloy) [reply src/main/java/seedu/address/model/entity/shop/Shop.java:1050]: refactored

File: src/main/java/seedu/address/logic/commands/AddServiceCommand.java [hunk 6]
```diff
@@ -53,21 +65,13 @@ public AddServiceCommand(Service service) {
...
-        if (!model.hasVehicle(toAdd.getVehicleId())) {
-            IdGenerator.setServiceIdUnused(toAdd.getId());
-            throw new CommandException(String.format(MESSAGE_VEHICLE_NOT_FOUND, toAdd.getVehicleId()));
-        }
-        if (model.hasService(toAdd.getId())) {
-            throw new CommandException(String.format(MESSAGE_DUPLICATE_SERVICE, toAdd.getId()));
+        try {
+            model.getShop().addService(vehicleId, entryDate, description,
+                    estimatedFinishDate, serviceStatus);
+            model.selectService(lst -> lst.get(lst.size() - 1));
```
[the remark this answers is not in the document]
Comment (ginloy) [reply src/main/java/seedu/address/logic/commands/AddServiceCommand.java:71]: I kinda lazy change now might change later

File: src/main/java/seedu/address/model/entity/shop/Shop.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (kimberlybp) [reply src/main/java/seedu/address/model/entity/shop/Shop.java:227]: yea i think the og code had 2 hasVehicle methods also 1 to compare id and another 1 to compare Vehicles which would then compare car plate numbers

image

iirc the one that compares car plate numbers is the validation for addVehicle command

File: src/main/java/seedu/address/model/entity/shop/exception/InsufficientPartException.java
Comment (kimberlybp) [reply src/main/java/seedu/address/model/entity/shop/exception/InsufficientPartException.java:1]: probably need to update/fix according to the shop updates :'')

**Comment by ginloy** [comment]
### Fix undo/redo
- Shifted logic (undo/redo stacks) over to Shop
- Added DeepCopy interface to ensure everything is deeply copied
- Got rid of some unused files
- Note: @kimberlybp I think there's still the bug where the panels attempt to update before maps and selected is updated when I undo but its still works so

**Pull request merged**